lib/evaluate/fold.cc

```diff
@@ -156,35 +162,298 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
     }
   }
   if (auto *intrinsic{std::get_if<SpecificIntrinsic>(&funcRef.proc().u)}) {
-    std::string name{intrinsic->name};
+    const std::string name{intrinsic->name};
     if (name == "kind") {
       if constexpr (common::HasMember<T, IntegerTypes>) {
-        return Expr<T>{args[0]->value->GetType()->kind};
+        return Expr<T>{funcRef.arguments()[0]->value->GetType()->kind};
       } else {
         common::die("kind() result not integral");
       }
     } else if (name == "len") {
       if constexpr (std::is_same_v<T, SubscriptInteger>) {
-        if (auto *charExpr{UnwrapExpr<Expr<SomeCharacter>>(*args[0]->value)}) {
+        if (auto *charExpr{UnwrapExpr<Expr<SomeCharacter>>(
+                *funcRef.arguments()[0]->value)}) {
           return std::visit([](auto &kx) { return kx.LEN(); }, charExpr->u);
         }
       } else {
         common::die("len() result not SubscriptInteger");
       }
-    } else {
-      // TODO: many more intrinsic functions
+    } else if (name == "iand") {
+      if (auto *x{std::get_if<BOZLiteralConstant>(
+              &funcRef.arguments()[0]->value->u)}) {
+        *funcRef.arguments()[0]->value =
+            Fold(context, ConvertToType<T>(std::move(*x)));
+      }
+      if (auto *x{std::get_if<BOZLiteralConstant>(
+              &funcRef.arguments()[1]->value->u)}) {
+        *funcRef.arguments()[1]->value =
+            Fold(context, ConvertToType<T>(std::move(*x)));
+      }
+      return FoldElementalIntrinsic<T, T, T>(
+          context, std::move(funcRef), ScalarFunc<T, T, T>(&Scalar<T>::IAND));
+    } else if (name == "int") {
+      return std::visit(
+          [&](auto &&x) -> Expr<T> {
+            using From = std::decay_t<decltype(x)>;
+            if constexpr (std::is_same_v<From, BOZLiteralConstant> ||
+                std::is_same_v<From, Expr<SomeReal>> ||
+                std::is_same_v<From, Expr<SomeInteger>> ||
+                std::is_same_v<From, Expr<SomeComplex>>) {
+              return Fold(context, ConvertToType<T>(std::move(x)));
+            } else {
+              common::die("int() argument type not valid");
+              return Expr<T>{std::move(funcRef)};  // unreachable
+            }
+          },
+          std::move(funcRef.arguments()[0]->value->u));
+    }
+    // TODO: many more intrinsic functions
+  }
+  return Expr<T>{std::move(funcRef)};
+}
+
+template<int KIND>
+Expr<Type<TypeCategory::Real, KIND>> FoldOperation(FoldingContext &context,
+    FunctionRef<Type<TypeCategory::Real, KIND>> &&funcRef) {
+  using namespace intrinsicHelper;
+  using T = Type<TypeCategory::Real, KIND>;
+  for (std::optional<ActualArgument> &arg : funcRef.arguments()) {
+    if (arg.has_value()) {
+      *arg->value = FoldOperation(context, std::move(*arg->value));
+    }
+  }
+  if (auto *intrinsic{std::get_if<SpecificIntrinsic>(&funcRef.proc().u)}) {
+    const std::string name{intrinsic->name};
+    if (name == "acos" || name == "acosh") {
```

Comment:
Is there eventually going to be a branch in this if-statement for every real intrinsic? It would be nice to have a way to break up this function so it doesn't end up being thousands of lines long.

Reply:
A way could be to handle all the cases that do not use host runtime (this will still be around 30/40 func for integers) and then to have a single call to the host runtime to try to fold it regardless the intrinsics.  The HostRte interface could be simplified into something like:
`hostRte.TryFolding(context, std::move(funcRef));`
This requires some work in HostRte and maybe a few duplicates with fold.cc, but it is probably cleaner. I think it is also possible to make this part independent of the actual intrinsics. This would make it easier to extend. I will describe my idea better before implementing it.